pkg/osvscanner/vulnerability_result.go

```diff
@@ -47,8 +51,45 @@ func groupResponseBySource(r *output.Reporter, query osv.BatchedQuery, resp *osv
 
 		pkg.Vulnerabilities = response.Vulns
 
-		pkg.Groups = grouper.Group(grouper.ConvertVulnerabilityToIDAliases(pkg.Vulnerabilities))
+		if groupsByIDBySource[query.Source] == nil {
+			groupsByIDBySource[query.Source] = make(map[string]*models.GroupInfo)
+		}
+
+		groups := grouper.Group(grouper.ConvertVulnerabilityToIDAliases(pkg.Vulnerabilities))
+		// TODO(rex): Is there a better way to do this, this feels fragile
+		for i := range groups {
+			for _, id := range groups[i].IDs {
+				groupsByIDBySource[query.Source][id] = &groups[i]
+			}
+		}
+		pkg.Groups = groups
 		groupedBySource[query.Source] = append(groupedBySource[query.Source], pkg)
+		vulnBySource[query.Source] = append(vulnBySource[query.Source], pkg.Vulnerabilities...)
+	}
+
+	if callAnalysis {
+		for si, v := range vulnBySource {
```

Comment:
Can we put this into its own function?

Reply:
The code is also getting a little hard to follow here with many interdependencies. Can we maybe simplify this all by doing something like this:

(and reusing the existing loop for putting together all the final details): 
```go

// from lines 95
for source, packages := range groupedBySource {
  // Do final source analysis to annotate output with call graph analysis results. 
  doSourceAnalysis(source, packages)
  output.Results = ...
}

// Maybe these should go a separate file as well

type MatchedVulnerability struct {
  v *Vulnerability
  group *models.GroupInfo
}

func vulnsFromAllPkgs(pkgs) []MatchedVulnerability,  {
   // collect all vulnerabilities and the matching groups 
}

func doSourceAnalysis(source, pkgs) {
  allVulns := vulnsFromAllPkgs(pkgs)
  // Maybe change the type of the second arg for this as well to `MatchedVulnerability`, since this step always acts as a filter/post-processing step after we match vulns 
  res, err := govulncheckshim.RunGoVulnCheck(filepath.Dir(source.Path.Path), allVulns)
  ...
   
}

```

Reply:
Did a refactor of this and the logic is split up nicely now, though I don't quite get what MatchedVulnerabilities is for. PTAL

Reply:
Thanks! Looks a lot cleaner now :)

`MatchedVulnerability` was just an abstraction to avoid returning a separate map of vulns to groupinfos. It also seemed like it made logical sense to me because it represented this idea of a "matched" vulnerability with additional context -- context that we add during post-scanning analysis (i.e. grouping, source analysis)

There may also be an edge case that `MatchedVulnerability` would help with. A single vulnerability may affect multiple different packages. If there are multiple matches, then the current `map[string]*models.GroupInfo` keyed on vuln ID may not be enough to disambiguate those cases right?

Reply:
I'm not clear on how having `MatchedVulnerabiliity` will help with the edge case, as we still need to store/index by package information to identify which package the analysis result is for right?

Reply:
The idea was that `MatchedVulnerability` can have duplicate vulns in there with different group mappings. 

The alternative here is for your map to have a slice of GroupInfos as values instead of a single one, which can also enable the fast lookup.